File: runtime-go/internal/encoder/security/zeek.go
package security

import (
	"fmt"

	"github.com/linxinhong/eventweave/runtime-go/internal/encoder"
	"github.com/linxinhong/eventweave/runtime-go/internal/event"
)

// ZeekConn encodes events as Zeek conn.log TSV entries.
type ZeekConn struct{}

func init() {
	encoder.Register("zeek-conn", ZeekConn{})
}

// Name returns the encoder name.
func (ZeekConn) Name() string { return "zeek-conn" }

// ContentType returns the MIME type of the encoded output.
func (ZeekConn) ContentType() string { return "text/tab-separated-values" }

// Encode formats the event as a Zeek conn.log tab-separated line.
func (ZeekConn) Encode(ev event.Event) ([]byte, error) {
	if err := require(ev, []string{"uid", "id.orig_h", "id.orig_p", "id.resp_h", "id.resp_p", "proto"}); err != nil {
		return nil, err
	}
	line := fmt.Sprintf(
		"%s\t%s\t%s\t%s\t%s\t%s\t%s",
		asString(attr(ev, "ts", ev.EventTime.Format("2006-01-02T15:04:05.000000Z"))),
		asString(ev.Attributes["uid"]),
		asString(ev.Attributes["id.orig_h"]),
		asString(ev.Attributes["id.orig_p"]),
		asString(ev.Attributes["id.resp_h"]),
		asString(ev.Attributes["id.resp_p"]),
		asString(ev.Attributes["proto"]),
	)
	return []byte(line), nil
}

// ZeekDNS encodes events as Zeek dns.log TSV entries.
type ZeekDNS struct{}

func init() {
	encoder.Register("zeek-dns", ZeekDNS{})
}

// Name returns the encoder name.
func (ZeekDNS) Name() string { return "zeek-dns" }

// ContentType returns the MIME type of the encoded output.
func (ZeekDNS) ContentType() string { return "text/tab-separated-values" }

// Encode formats the event as a Zeek dns.log tab-separated line.
func (ZeekDNS) Encode(ev event.Event) ([]byte, error) {
	if err := require(ev, []string{"uid", "id.orig_h", "id.orig_p", "id.resp_h", "id.resp_p", "query"}); err != nil {
		return nil, err
	}
	line := fmt.Sprintf(
		"%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s",
		asString(attr(ev, "ts", ev.EventTime.Format("2006-01-02T15:04:05.000000Z"))),
		asString(ev.Attributes["uid"]),
		asString(ev.Attributes["id.orig_h"]),
		asString(ev.Attributes["id.orig_p"]),
		asString(ev.Attributes["id.resp_h"]),
		asString(ev.Attributes["id.resp_p"]),
		asString(ev.Attributes["query"]),
		asString(attr(ev, "qtype_name", "")),
	)
	return []byte(line), nil
}

File: packs/security/encoders/zeek.py
"""Zeek conn.log and dns.log encoders."""

from __future__ import annotations

from eventweave.core.event import Event
from eventweave.encoders import Encoder, EncodeResult, encoder
from packs.security.encoders._helpers import event_time_str, format_tsv


@encoder("zeek-conn", content_type="text/plain")
class ZeekConnEncoder(Encoder):
    """Encode events as Zeek conn.log TSV records."""

    name = "zeek-conn"
    content_type = "text/plain"
    description = "Zeek conn.log TSV network connection format."
    required_fields = ["uid", "id.orig_h", "id.orig_p", "id.resp_h", "id.resp_p", "proto"]
    optional_fields = [
        "service",
        "duration",
        "orig_bytes",
        "resp_bytes",
        "conn_state",
    ]
    supported_event_types = ["network.connection"]
    _columns = [
        "ts",
        "uid",
        "id.orig_h",
        "id.orig_p",
        "id.resp_h",
        "id.resp_p",
        "proto",
        "service",
        "duration",
        "orig_bytes",
        "resp_bytes",
        "conn_state",
    ]

    def encode(self, event: Event) -> EncodeResult:
        missing = [f for f in self.required_fields if f not in event.attributes]
        if missing:
            return self._fail(f"missing required fields: {', '.join(missing)}")

        values: list[object] = [event_time_str(event)]
        for col in self._columns[1:]:
            values.append(event.attributes.get(col))
        return self._ok(format_tsv(values))


@encoder("zeek-dns", content_type="text/plain")
class ZeekDNSEncoder(Encoder):
    """Encode events as Zeek dns.log TSV records."""

    name = "zeek-dns"
    content_type = "text/plain"
    description = "Zeek dns.log TSV DNS query/response format."
    required_fields = ["uid", "id.orig_h", "id.orig_p", "id.resp_h", "id.resp_p", "query"]
    optional_fields = [
        "proto",
        "trans_id",
        "qtype_name",
        "rcode_name",
        "answers",
    ]
    supported_event_types = ["network.dns"]
    _columns = [
        "ts",
        "uid",
        "id.orig_h",
        "id.orig_p",
        "id.resp_h",
        "id.resp_p",
        "proto",
        "trans_id",
        "query",
        "qtype_name",
        "rcode_name",
        "answers",
    ]

    def encode(self, event: Event) -> EncodeResult:
        missing = [f for f in self.required_fields if f not in event.attributes]
        if missing:
            return self._fail(f"missing required fields: {', '.join(missing)}")

        values: list[object] = [event_time_str(event)]
        for col in self._columns[1:]:
            value = event.attributes.get(col)
            if col == "answers" and isinstance(value, list):
                value = ",".join(str(v) for v in value)
            values.append(value)
        return self._ok(format_tsv(values))
